File: crates/tvm-core/src/policy.rs
use crate::region::RegionKind;
use crate::residency::Residency;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct PlacementPolicy {
    pub initial_residency: Residency,
    pub pinnable: bool,
    pub spillable: bool,
}

impl PlacementPolicy {
    pub const fn for_kind(kind: RegionKind) -> Self {
        match kind {
            RegionKind::HotHeap | RegionKind::CodeCache | RegionKind::DeviceState => Self {
                initial_residency: Residency::Hot,
                pinnable: true,
                spillable: false,
            },
            RegionKind::ObjectArena | RegionKind::BlobArena => Self {
                initial_residency: Residency::Hot,
                pinnable: false,
                spillable: true,
            },
            RegionKind::PageStore => Self {
                initial_residency: Residency::Warm,
                pinnable: false,
                spillable: true,
            },
            RegionKind::Scratch => Self {
                initial_residency: Residency::Hot,
                pinnable: false,
                spillable: false,
            },
        }
    }
}

File: web/tvm-web/src/policy.ts
/** Port of `tvm-core::PlacementPolicy` (crates/tvm-core/src/policy.rs). */
import { RegionKind, Residency } from "./types.js";

export interface PlacementPolicy {
  initialResidency: Residency;
  pinnable: boolean;
  spillable: boolean;
}

export function policyForKind(kind: RegionKind): PlacementPolicy {
  switch (kind) {
    case RegionKind.HotHeap:
    case RegionKind.CodeCache:
    case RegionKind.DeviceState:
      return { initialResidency: Residency.Hot, pinnable: true, spillable: false };
    case RegionKind.ObjectArena:
    case RegionKind.BlobArena:
      return { initialResidency: Residency.Hot, pinnable: false, spillable: true };
    case RegionKind.PageStore:
      return { initialResidency: Residency.Warm, pinnable: false, spillable: true };
    case RegionKind.Scratch:
      return { initialResidency: Residency.Hot, pinnable: false, spillable: false };
  }
}
